qiskit/qpy/binary_io/circuits.py

```diff
@@ -41,11 +46,81 @@
 from qiskit.circuit.quantumcircuit import QuantumCircuit
 from qiskit.circuit import QuantumRegister, Qubit
 from qiskit.qpy import common, formats, type_keys
+from qiskit.qpy.exceptions import QpyError, UnsupportedFeatureForVersion
 from qiskit.qpy.binary_io import value, schedules
 from qiskit.quantum_info.operators import SparsePauliOp, Clifford
 from qiskit.synthesis import evolution as evo_synth
 from qiskit.transpiler.layout import Layout, TranspileLayout
 
+if typing.TYPE_CHECKING:
+    from qiskit.circuit.annotation import QPYSerializer, Annotation
+
+
+class _AnnotationSerializationState:
+    def __init__(self, factories: dict[str, typing.Callable[[], QPYSerializer]]):
+        self.factories = factories
+        self.serializers = {}
+        self.potential_serializers = {}
+
+    @property
+    def num_serializers(self) -> int:
+        """The number of constructed serializers."""
+        return len(self.serializers)
+
+    def serialize(self, annotation: Annotation) -> (int, bytes):
+        """Serialize an annotation using a known serializer (initializing one, if necessary).
+
+        Returns the index of the serializer used, and the serialized annotation."""
+        for namespace in iter_namespaces(annotation.namespace):
+            if (existing := self.serializers.get(namespace, None)) is not None:
+                index, serializer = existing
+                if (out := serializer.dump_annotation(annotation)) is not NotImplemented:
+                    return index, out
+            if (serializer := self.potential_serializers.get(namespace, None)) is not None:
+                if (out := serializer.dump_annotation(annotation)) is not NotImplemented:
+                    del self.potential_serializers[namespace]
+                    index = len(self.serializers)
```

Comment:
Should we have size checking on the number of serializers to make sure we don't exceed the width of a `u32`? I don't think we do that for anything else in QPY though, anyway.

Reply:
that's like 4 billion or something - I think you'd have larger problems if you were exceeding that